### DISCUSSION TITLE
Need help - BF not working

### DISCUSSION THREAD

**Discussion by KelsonA** [neuropoly/axondeepseg#748] [Help needed]
Hi my name is Kelson, I'm a student researcher very new to this whole program. I've followed the installation guide and seem to have napari and the axondeepseg plugin working. When I run one of our images using the BF rat axon model, I get the following error traceback (below):

I dont really know what I'm doing and could really use some help with getting this running as well as any recommendations you have for optimizing settings, etc.

I can run these same images through the SEM model, though the results are awful as these aren't EM images, just thought Id include that incase it signifies I have installed everything correctly.

For context Im running an M1 MacBook Pro - and I installed everything through python, not the simple package installation. 

Thanks so much for your help! 

---------------------------------------------------------------------------
SystemExit                                Traceback (most recent call last)
File ~/axondeepseg/plugins/napari_ADS/_widget.py:844, in ApplyModelThread.run(self=<plugins.napari_ADS._widget.ApplyModelThread object>)
    843 try:
--> 844     segment.segment_image(
        segment = <module 'AxonDeepSeg.segment' from '/Users/kelsonadcock/axondeepseg/AxonDeepSeg/segment.py'>
        self = <plugins.napari_ADS._widget.ApplyModelThread object at 0x281533e50>
        self.path_testing_image = PosixPath('/Users/kelsonadcock/Desktop/Axondeepseg/DownloadsNKB/image.png')
        self.path_model = PosixPath('/Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf')
        self.overlap_value = [48, 48]
        self.selected_layer = <Image layer 'image' at 0x1739dd400>
        self.zoom_factor = 1.0
        self.gpu_id = 0
        self.no_patch = False
    845         path_testing_image=self.path_testing_image,
    846         path_model=self.path_model,
    847         overlap_value=self.overlap_value,
    848         acquired_resolution=self.selected_layer.metadata["pixel_size"],
    849         zoom_factor=self.zoom_factor,
    850         gpu_id=self.gpu_id,
    851         no_patch=self.no_patch,
    852         verbosity_level=3,
    853     )
    854     self.task_finished_successfully = True

File ~/miniconda3/envs/ads_venv/lib/python3.8/site-packages/loguru/_logger.py:1251, in Logger.catch.<locals>.Catcher.__call__.<locals>.catch_wrapper(*args=(), **kwargs={'acquired_resolution': 0.07, 'gpu_id': 0, 'no_patch': False, 'overlap_value': [48, 48], 'path_model': PosixPath('/Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf'), 'path_testing_image': PosixPath('/Users/kelsonadcock/Desktop/Axondeepseg/DownloadsNKB/image.png'), 'verbosity_level': 3, 'zoom_factor': 1.0})
   1250 with catcher:
-> 1251     return function(*args, **kwargs)
        function = <function segment_image at 0x28098bd30>
        args = ()
        kwargs = {'path_testing_image': PosixPath('/Users/kelsonadcock/Desktop/Axondeepseg/DownloadsNKB/image.png'), 'path_model': PosixPath('/Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf'), 'overlap_value': [48, 48], 'acquired_resolution': 0.07, 'zoom_factor': 1.0, 'gpu_id': 0, 'no_patch': False, 'verbosity_level': 3}
   1252 return default

File ~/axondeepseg/AxonDeepSeg/segment.py:192, in segment_image(path_testing_image=PosixPath('/Users/kelsonadcock/Desktop/Axondeepseg/DownloadsNKB/image.png'), path_model=PosixPath('/Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf'), overlap_value=[48, 48], acquired_resolution=0.07, zoom_factor=1.0, no_patch=False, gpu_id=0, verbosity_level=3)
    191     logger.error(error_msg)
--> 192     sys.exit(4)
    194 # Performing the segmentation

SystemExit: 4

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
File ~/axondeepseg/plugins/napari_ADS/_widget.py:857, in ApplyModelThread.run(self=<plugins.napari_ADS._widget.ApplyModelThread object>)
    855 except SystemExit as err:
    856     if err.code == 4:
--> 857         self.show_info_message(
        self = <plugins.napari_ADS._widget.ApplyModelThread object at 0x281533e50>
    858             "Resampled image smaller than model's patch size. Please take a look at your terminal \n"
    859             "for the minimum zoom factor value to use (option available in the Settings menu)."
    860         )
    861     self.task_finished_successfully = False
    862 self.model_applied_signal.emit()

AttributeError: 'ApplyModelThread' object has no attribute 'show_info_message'

**Comment by Stoyan-I-A** [comment]
Hello @KelsonA , thank you for reaching out. I noticed thanks to your log output that there's a bug where some information should be displayed but isn't.

The information that should have been displayed is  `Resampled image smaller than model's patch size. Please take a look at your terminal for the minimum zoom factor value to use (option available in the Settings menu).`
What is the pixel size of your image (that you either entered in napari or have as information in a `pixel_size_in_micrometer.txt` file)? And are the dimensions of your image?

If you want, you can share a sample of the image just so that we know what kind of data you're trying to segment.

**Reply by KelsonA** [reply]
OK, so I was able to get BF to run by changing the zoom variable. However, I think my settings are still very unoptimized. Ive done many runs changing the zoom value between 1.45 (anything much lower than this won't run), and 2.1. It seems by changing this zoom value I pick up some axons and lose others? doesn't seem like a great way to optimize my settings. Would anyone have any other ideas as to how I can better highlight more of the axons in my slides?
<img width="382" alt="Screen Shot 2023-06-23 at 1 03 47 PM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/51a80767-80c5-4068-b214-4071842b87a4">

**Reply by KelsonA** [reply]
Hi Stoyan, sorry I hadnt seen your reply before I replied to my own comment. above is one of the images Im working with, its a cross section of a rat nerve. I havent messed with the pixel size of my image variable, as to be honest I have no idea what that is or would be. Ive left it at the default .7 value. I simply changed the zoom around in order to get the BF model to run, and it seems I can change the number of axons it picks up by changing to zoom value to some degree, but as you can see I'm still missing a lot of axons. Any ideas how I can optimize the settings to pick up more of them?

**Reply by KelsonA** [reply]
If I right click on my .png file and hit get info - it says the dimensions are 510 x 618, is that the pixel size? what would I input into hat setting as a single variable in that case?

**Reply by Stoyan-I-A** [reply]
So 510 x 618 are the dimensions of the image, not the pixel size. The pixel size corresponds to how long a pixel is terms of "real world" or "physical" units. For exemple, a pixel size of 0.7 means that every pixel of the image correspond to a distance of 0.7 micrometers in "real life". 
This information is determined when the images are acquired. We can also deduce it if a scale bar is on the image.

Now, using the right pixel size doesn't guarantee that the automatic segmentation outputs the most optimal result. Which is why we added the zoom factor. Your pixel size will be multiplied by the zoom factor before applying the automatic segmentation.

Obtaining the most optimal result will require a bit of trial and error (adjusting the pixel size until you get the best segmentation possible with our models). 
If you want to use the command line, there is an option to specify a range of zoom factors to try out: https://axondeepseg.readthedocs.io/en/latest/documentation.html#segment-an-image-using-a-range-of-zoom-factors

**Reply by KelsonA** [reply]
So, I contacted another lab mate who took the photos of our slides. They said based on the objective of the lens they used, that we have a "microns per pixel value of .16125". would the pixels per micron be the inverse of that then at 6.202?

**Reply by Stoyan-I-A** [reply]
In this case, your pixel size would be 0.16125 (micrometers per pixel)

**Reply by KelsonA** [reply]
OK, so I've run this through a number of times with .16125 as the pixel size while varying the zoom value (bottom right corner of these screenshots has (.16125, zoom)). Anything much lower than .7 zoom value will fail to run, and above 1.1 many axons are omitted. Are these images what you would expect in terms of the number of axons the plugin is able to capture? Are there any other settings I could try tweaking to capture more of the axons? I would assume with a clearer image (there is a band in this one that is quite blurry where nothing is picked up by the software), perhaps axon capture might improve? There are also some areas/axons that seem quite clear that it doesn't seem to pick up on. I guess any advice you might have for me at this point would be useful! I want to get an idea of what consistent methods I can apply going forward before I start compiling morphometrics for our study, so any advice you might have at this stage would be invaluable before we move forward!

**Reply by KelsonA** [reply]
<img width="343" alt="Screen Shot 2023-06-24 at 11 23 28 AM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/3aa8451f-5755-4037-a11a-b52be5ba73dc">

**Reply by KelsonA** [reply]
<img width="348" alt="Screen Shot 2023-06-24 at 11 27 50 AM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/db36fce3-312f-48e5-9915-968b89ab903b">

**Reply by KelsonA** [reply]
<img width="347" alt="Screen Shot 2023-06-24 at 11 22 26 AM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/af436858-6d8f-4fc4-883e-795d6aae660a">

**Reply by KelsonA** [reply]
<img width="345" alt="Screen Shot 2023-06-24 at 10 56 03 AM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/383ea08b-acf0-464a-9646-669ded7c8006">

**Reply by KelsonA** [reply]
<img width="348" alt="Screen Shot 2023-06-24 at 11 20 08 AM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/96094078-2392-4ea7-9737-920d0e23e2ec">

**Reply by KelsonA** [reply]
<img width="347" alt="Screen Shot 2023-06-24 at 10 57 51 AM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/805603a2-acda-4bcf-9b0c-407872918652">

**Reply by Stoyan-I-A** [reply]
For further advice on obtaining better segmentations, I'm going to tag @hermancollin and @mathieuboudreau if they have any ideas.
The best way to obtain a good segmentation would be to train a new model, but this is only worth doing if you have many more images. If you don't have too many images, you can start with the segmentation output from AxonDeepSeg and correct the rest manually.

Also for your information, the reason why zoom factors lower than ~1.1~ 0.7 will fail with your data is because your images are small compared to the patch size used to train the original bf model. The original model used patches of dimensions 512 x 512, and your dimensions are 510 x 618. So if your pixel size is similar to the one used for training the model (I unfortunately do not recall what it was), a zoom factor of 1.0 will be unable to segment your images since your lowest dimension is 510. The images used to train the BF model were quite large (above 2000 x 2000).

**Reply by KelsonA** [reply]
Ah that makes sense! If Herman or Mathieu have any other ideas for obtaining better segmentations, that would be awesome! We have a good number of images, but idk if we have enough to train a new model, or how to go about doing that. I'll ask my PI tomorrow how many we have and get back to you. Is training a new model difficult? Im sure you can tell by now I dont have a lot of experience with python or any of this sort of tech, so unless there are simple instructions written out somewhere I probably wouldn't be able to manage. 

I will probably look to start running images for the study at the end of this coming week, so any tips on optimization until then would be greatly appreciated! Thank you for your help!

**Reply by Stoyan-I-A** [reply]
Our current models are trained using Ivadomed (documentation [here](https://ivadomed.org/tutorials/two_class_microscopy_seg_2d_unet.html)). 
A possible problem that I can anticipate is that if you have any issues with the training, there may not be anyone to help because work on the Ivadomed project has been stopped. 
We want to move away from Ivadomed for this reason, but this will take a few months of development on our end.

**Reply by KelsonA** [reply]
were hoping to move forward and have some preliminary data sooner than that, so we will probably stick with what we've got for now, and do our best to optimize this model as best we can. If @hermancollin or @mathieuboudreau have any ideas on optimization settings I can try wed be very grateful for any advice! Thanks again!

**Reply by mathieuboudreau** [reply]
Hi @KelsonA - could you possibly send us one of your histology images so that we can see if we can improve on your segmentation attempt? You can email it to axondeepseg@googlegroups.com

**Reply by KelsonA** [reply]
Just sent over 2 slides - 1 that we've been referencing in this post, and another that has larger dimensions (in case that's potentially the culprit), and is representative of many of our other slides. Thank you again!

**Reply by KelsonA** [reply]
pixel size should be .16125 microns per pixel for both as they were taken with the same objective lens

**Comment by mathieuboudreau** [comment]
@KelsonA I received your two sample images, thank you for sending them!

Last night I took a closer look at them both, and explored different settings in ADS to see how well that the automated segmentation could perform on them.

Before going into more technical details, here's my opinion on your situation

* Your images have a low resolution (i.e. number of pixels per axon), and this likely is the root cause why a lot of the smaller axons are being "missed" in the segmentation.
  * Potential solutions:
    * Save your images in a higher quality image format/resolution at the microscope
    * Instead of saving one image of the entire fibre / region of interest, if possible use a higher zoom lens at the microscope and save those images (i.e. instead of one low resolution image, you'll have multiple segments that are higher resolution)
* Image artefacts (like some "smearing" seen in your images) are not things that AxonDeepSeg could overcome on it's own
  * Potential solutions:
    * It's likely that you would need to manually segment these.
    * Talk with the person preparing your samples to see what the root cause of these might be, and how this could be avoided in future samples
* Low image contrast (the second image you send, where everything is very dark) is not the best condition for automatic segmentation tools (AxonDeepSeg or other ones).
  * Potential solutions:
    * Adjust the light/brightness settings on your microscope in future sample collection, if possible
    * For images that have already been saved, you could improve performance by adjusting the image properties (eg brightness, contrast, etc) in any image editing tools (like Lightroom, or Apple's Photo app)


------------


**For your first image,**

![image copy](https://github.com/axondeepseg/axondeepseg/assets/1421029/c312525f-b72c-45a0-a96e-082faa348d4c)

the first thing that I noticed was that it is quite a low resolution - I could see the "pixelation" at the myelin-axon boundary by looking at the image at full scale (without even zooming in).


<img width="698" alt="Screenshot 2023-06-27 at 10 42 33 AM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/de103fc8-21b7-41ca-a2bb-58abfea6aaa3">

Using an image of this size (510 × 618) for this many axons means that there will be less information for the deep learning model to use to "identify" what's an axon and what isn't. Another problem is that there is some image resampling being done at some steps of the segmentation (also if you use the zoom factor, this is what's being done), so the smaller axons could get "lost" during this resampling. 

The best I could achieve on this image was the same as what you did, with the BF model / zoom=1 / pixel size =0.16125,

<img width="636" alt="Screenshot 2023-06-26 at 9 11 35 PM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/7609687b-2093-4922-b634-f3560200318d">

Using an additional option (no patches) improved some of the bigger axons, but lost some of the smaller ones,

<img width="636" alt="Screenshot 2023-06-26 at 9 12 49 PM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/1e424c0b-d09b-4afe-bde6-29990c35f196">

I also tried using the TEM model on this image (with zoom factor of ~0.2), as this model uses images of similar contrast as BF and sometimes performs well with these images, but found that BF still had better results. TEM resulted in some areas outside the fiber being segmented, so more cleanup would be necessary,

<img width="636" alt="tem_0p2z" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/e9700f27-ade0-4eda-88d5-e808a62788ca">


**For your second image,**


![59L SLN v10BFvis](https://github.com/axondeepseg/axondeepseg/assets/1421029/a923af46-37e2-4179-a4ec-b0a358c39055)

it has *very* low contrast (ie. the bright areas are about as dark as the darkest areas). I did not find any settings that worked well at segmenting these images as-is,

<img width="811" alt="Screenshot 2023-06-27 at 10 47 40 AM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/77ef3236-d4b1-4a14-a79d-d4262ba756e3">

I explored how well I could improve the contrast via image post-processing; I used my Mac's "Photos" software, here's how well I could improved it and the main settings changed,

<img width="1194" alt="Screenshot 2023-06-26 at 9 17 37 PM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/23e08daa-91f7-46fe-8a6c-732301f4cb19">

Using this updated image, I didn't get good results with the BF model, but I did get decent results using the TEM model (likely because I artificially changed the image contrast to a way that it looks more like a TEM image than BF image?),

<img width="846" alt="Screenshot 2023-06-26 at 9 26 34 PM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/5813f6fe-590c-4f7f-84f5-93b1afef5ab3">

Like the previous image, there are some parts outside the axon that did get incorrectly segmented that would need to be cleaned up, but a lot more axons were segmented. If you had a lot of images with exactly this contrast, I know lots of tools like Lightroom allow for applying the same image settings on a collection of images, but I'd recommend in the future to improve the image contrast at the source (microscope).

See the first part of this post for some of my insights/advice.

All-in-all, I don't think this is an issue of the BF model not working correctly, but that your images are ill-suited to be segmented using automated tools due to their low resolution, low contrast, and image artefacts.

**Reply by KelsonA** [reply]
Hey thank you so much for all this information! I feel good about moving forward and will chat with my PI about potential methods we could implement for determining a consistent approach to manually selecting some of the remaining axons! Thanks!

**Reply by mathieuboudreau** [reply]
On another note, following some an idea from @hermancollin during another meeting today, I tried another strategy on that last image, where inverted the contrast and then tried our SEM model to segment it - it performed better I think! Settings: SEM model, true pixel size, zoom factor 0.4, no-patch option on.

<img width="1440" alt="Screenshot 2023-06-28 at 6 56 55 PM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/2c31b271-5aae-4df5-b66a-366459bfbb05">

The reasoning for this strategy is that the SEM model was trained using more "dense" images (lots of axons in a single image), whereas TEM wasn't (few axons in those images). The contrast between TEM and SEM is somewhat an inversion between the two, so that step is necessary. I find it segmented a lot of the little axons better.

**Comment by KelsonA** [comment]
One more question if you dont mind - I've been running some other files, and have come up against an issue - sometimes when I load in an image (it SEEMS to me like its more common with .tif files, but im not sure), it changes the settings from the standard "RGB" Ive been dealing with up till now to a "grey" option. when I bring down the menu to try and find RGB to select the option isn't there, and this results in there being no color overlay (the red and blue). Or maybe just none of the axons are getting picked up at all? Any ideas what might be going on an what I can do to troubleshoot? Its happened with a pretty significant number of my images. Ill attach a few screenshots for reference

**Reply by KelsonA** [reply]
<img width="1121" alt="Screen Shot 2023-06-28 at 2 16 44 PM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/5a175940-ed37-4790-9d44-151ea52669f1">

this one is the normal and the add-on runs fine and gives the red/blue overlay

**Reply by KelsonA** [reply]
<img width="1117" alt="Screen Shot 2023-06-28 at 2 21 45 PM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/ee214cb5-d176-44a7-95f2-d7892b66567b">

**Reply by KelsonA** [reply]
<img width="1119" alt="Screen Shot 2023-06-28 at 2 16 31 PM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/97ed4f1c-cd55-4b10-9053-5ce2c8015f32">

**Reply by KelsonA** [reply]
here you can see the colormap: defaults to "gray", with no RGB option to be found. And then you can see after running it, either the color isn't showing up (that was just my guess at the issue due to the gray option), or perhaps it isn't running at all as the image looks like it hasn't had anything overlayed? Any ideas? Thanks!

**Reply by mathieuboudreau** [reply]
Hum, this sounds more like a napari software question and not about our plugin, but maybe @Stoyan-I-A has some insights?

**Reply by KelsonA** [reply]
seems to happen to every .tif, the add-on runs, and 2 layers show up, but no changes are visible
<img width="1114" alt="Screen Shot 2023-06-28 at 3 29 43 PM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/be484866-9bbf-4a35-8f5e-cca73aeb33da">

**Reply by Stoyan-I-A** [reply]
@KelsonA can you copy and paste the output of the terminal when trying to segment a .tif image?

On another note, the segmentation is not exactly performed on the image displayed in napari. We pass the path (location) of the image that was loaded to the Ivadomed package, which is the package that performs the segmentation, and then load in napari the segmentation results (if it succeeded, they are located in the same folder as the original image).

**Reply by Stoyan-I-A** [reply]
Also check if the .tif image might be a 16 bit image (or even 32 bit). We had issues with those in the past.

**Reply by Stoyan-I-A** [reply]
But as for the reason why .tif images are loaded as RGB, this is on the napari side. I don't really have an answer. Maybe the image is encoded as RGB and can't be changed even though it looks greyscale.
But the fact that the segmentation is empty is something to investigate on the AxonDeepSeg or Ivadomed side. You could share a sample .tif image if you want.

**Reply by mathieuboudreau** [reply]
Yeah, likely a 16bit issue, see #737

**Reply by KelsonA** [reply]
yeah of course - here is the terminal readout

INFO: Applying ADS model... This can take a few seconds. Check the console for more information.
15:07:55 WARNING Lossy conversion from >u2 to uint8. Range [6125, 27604]. Convert image to uint8 prior to saving to suppress this warning.
2023-06-29 15:07:55.500 | WARNING  | AxonDeepSeg.apply_model:axon_segmentation:49 - The 'no-patch' option was not selected for segmentation. Please note that this option could potentially produce better results but may not be suitable with large images depending on computer RAM capacity.
2023-06-29 15:07:55.503 | INFO     | ivadomed.utils:define_device:160 - Cuda is not available.
2023-06-29 15:07:55.503 | INFO     | ivadomed.utils:define_device:161 - Working on cpu.
2023-06-29 15:07:55.504 | INFO     | ivadomed.config_manager:_display_differing_keys:202 - Adding the following keys to the configuration file
2023-06-29 15:07:55.504 | INFO     | ivadomed.config_manager:deep_dict_compare:47 -     wandb: {'wandb_api_key': '', 'project_name': 'my_project', 'group_name': 'my_group', 'run_name': 'run-1', 'log_grads_every': 100}
2023-06-29 15:07:55.504 | INFO     | ivadomed.config_manager:deep_dict_compare:47 -     loader_parameters: patch_filter_params: {'filter_empty_mask': False, 'filter_empty_input': False}
2023-06-29 15:07:55.505 | INFO     | ivadomed.config_manager:deep_dict_compare:47 -     evaluation_parameters: object_detection_metrics: True
2023-06-29 15:07:55.505 | INFO     | ivadomed.config_manager:_display_differing_keys:204 - 

2023-06-29 15:07:55.507 | WARNING  | ivadomed.inference:segment_volume:409 - fname_roi has not been specified, then the entire volume is processed.
2023-06-29 15:07:56.165 | INFO     | ivadomed.loader.mri2d_segmentation_dataset:determine_cache_need:378 - Estimated 2D dataset size is 3.4226104617118835e-06 GB, which is smaller than 4.257259368896484 GB. File cache will not be used
2023-06-29 15:07:56.178 | INFO     | ivadomed.inference:segment_volume:477 - Loaded 20 axial patches of shape [512, 512].
2023-06-29 15:07:56.229 | DEBUG    | ivadomed.inference:get_preds:87 - Likely ONNX model detected at: /Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf/model_seg_rat_axon-myelin_bf.onnx
2023-06-29 15:07:56.237 | DEBUG    | ivadomed.inference:get_preds:88 - Conduct ONNX model inference... 
2023-06-29 15:08:02.193 | DEBUG    | ivadomed.inference:get_preds:91 - Sending predictions to CPU
2023-06-29 15:08:02.215 | DEBUG    | ivadomed.inference:get_preds:87 - Likely ONNX model detected at: /Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf/model_seg_rat_axon-myelin_bf.onnx
2023-06-29 15:08:02.215 | DEBUG    | ivadomed.inference:get_preds:88 - Conduct ONNX model inference... 
2023-06-29 15:08:07.601 | DEBUG    | ivadomed.inference:get_preds:91 - Sending predictions to CPU
2023-06-29 15:08:07.615 | DEBUG    | ivadomed.inference:get_preds:87 - Likely ONNX model detected at: /Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf/model_seg_rat_axon-myelin_bf.onnx
2023-06-29 15:08:07.615 | DEBUG    | ivadomed.inference:get_preds:88 - Conduct ONNX model inference... 
2023-06-29 15:08:12.861 | DEBUG    | ivadomed.inference:get_preds:91 - Sending predictions to CPU
2023-06-29 15:08:12.875 | DEBUG    | ivadomed.inference:get_preds:87 - Likely ONNX model detected at: /Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf/model_seg_rat_axon-myelin_bf.onnx
2023-06-29 15:08:12.875 | DEBUG    | ivadomed.inference:get_preds:88 - Conduct ONNX model inference... 
2023-06-29 15:08:18.353 | DEBUG    | ivadomed.inference:get_preds:91 - Sending predictions to CPU
2023-06-29 15:08:18.366 | DEBUG    | ivadomed.inference:get_preds:87 - Likely ONNX model detected at: /Users/kelsonadcock/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_bf/model_seg_rat_axon-myelin_bf.onnx
2023-06-29 15:08:18.366 | DEBUG    | ivadomed.inference:get_preds:88 - Conduct ONNX model inference... 
2023-06-29 15:08:23.685 | DEBUG    | ivadomed.inference:get_preds:91 - Sending predictions to CPU
2023-06-29 15:08:23.873 | INFO     | AxonDeepSeg.segment:segment_image:201 - Image /Users/kelsonadcock/Desktop/59L SLN v10BFvis.tif segmented.

**Reply by KelsonA** [reply]
And yes, I will email that .tif file over right now

**Reply by KelsonA** [reply]
I'm unsure how to determine if its a 16 or 32 bit... is there info here?

<img width="397" alt="Screen Shot 2023-06-29 at 3 12 45 PM" src="https://github.com/axondeepseg/axondeepseg/assets/137534495/d0339067-4a11-4937-a251-8c1b723878c4">

**Reply by mathieuboudreau** [reply]
Yup it is 16 bit (see top of image)

<img width="1245" alt="Screenshot 2023-06-29 at 8 52 44 PM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/d0aa4abe-0c7d-4931-86c8-6214cfd53a48">

The issue is on the Ivadomed side, and I've already opened a pull request with a fix: https://github.com/ivadomed/ivadomed/pull/1296

In the meantime, if you want to segment it, all you need to do is convert it to 8 bits using an image editing tool like Gimp or Adobe Photoshop, or even by using the export tools like Preview on a Mac:

<img width="960" alt="Screenshot 2023-06-29 at 8 55 35 PM" src="https://github.com/axondeepseg/axondeepseg/assets/1421029/f56dd2e4-8e17-4a22-a71d-6010f863adee">

**Reply by KelsonA** [reply]
That's awesome! I exported a few to 8-bit and it worked wonders! I'll continue to try and figure out how to optimize things, thank you so much for answering so many disjointed questions so quickly simply and concisely. It has really been helpful!
